File: js/test/Exchange/test.watchOHLCV.js
'use strict'

// ----------------------------------------------------------------------------

const log = require ('ololog')
    , chai = require ('chai')
    , asTable = require ('as-table')
    , assert = chai.assert
    , testOHLCV = require ('ccxt/js/test/Exchange/test.ohlcv.js')

/*  ------------------------------------------------------------------------ */

module.exports = async (exchange, symbol) => {

    // log (symbol.green, 'watching ohlcv...')

    const method = 'watchOHLCV'

    if (!exchange.has[method]) {
        log (exchange.id, 'does not support', method + '() method')
        return
    }

    const timeframe = Object.keys (exchange.timeframes)[0]

    let response = undefined

    let now = Date.now ()
    const ends = now + 30000

    while (now < ends) {

        response = await exchange[method] (symbol, timeframe)

        now = Date.now ()

        assert (response instanceof Array)
        // log (symbol.green, method, 'returned', Object.values (response).length.toString ().green, 'ohlcvs')
        for (let j = 0; j < response.length; j++) {
            testOHLCV (exchange, response[j], symbol, now)
            if (j > 0) {
                if (response[j][0] && response[j - 1][0]) {
                    assert (response[j][0] >= response[j - 1][0])
                }
            }
        }

        response = response.map ((ohlcv) => [
            exchange.iso8601 (ohlcv[0]),
            symbol,
            ohlcv[1],
            ohlcv[2],
            ohlcv[3],
            ohlcv[4],
            ohlcv[5],
        ])

        console.log ('--------------------------------------------------------')
        log.noLocate (asTable (response))
    }

    return response
}

File: python/test/exchange/test_watch_ohlcv.py
# -*- coding: utf-8 -*-

from exchange.test_ohlcv import test_ohlcv
from ccxt import NetworkError
# from exchange.table import table


async def test_watch_ohlcv(exchange, symbol):
    # todo add real ohlcv tests here
    # log (symbol.green, 'watching ohlcvs...')
    method = 'watchOHLCV'

    skipped_exchanges = [
        'dsx',
        'idex2',  # rinkeby testnet, trades too rare
    ]

    if exchange.id in skipped_exchanges:
        print(exchange.id, method, 'skipped')
        return

    try:
        timeframe = list(exchange.timeframes.keys())[0]
    except Exception:
        timeframe = '1d'
    limit = 10
    duration = exchange.parse_timeframe(timeframe)
    since = exchange.milliseconds() - duration * limit * 1000 - 1000
    if (method in exchange.has) and exchange.has[method]:
        response = None
        now = exchange.milliseconds()
        end = now + 15000
        while now < end:
            try:
                ohlcvs = await getattr(exchange, method)(symbol, timeframe, since, limit)
                now = exchange.milliseconds()
                print(exchange.iso8601(now), symbol, timeframe, len(ohlcvs), 'ohlcvs')
                for ohlcv in ohlcvs:
                    test_ohlcv(exchange, ohlcv, method, symbol)
                # print(table([[exchange.iso8601(o[0])] + o[1:] for o in ohlcvs]))
            except NetworkError:
                now = exchange.milliseconds()
        return response
    else:
        print(exchange.id, method, 'is not supported or not implemented yet')


__all__ = ['test_watch_ohlcv']
